libr/anal/type.c

```diff
@@ -122,18 +122,119 @@ R_API void r_anal_save_parsed_type(RAnal *anal, const char *parsed) {
-		r_list_append (type_used , var->type);
-        }
-	RList *uniq = r_list_uniq (type_used , typecmp);
+	RList *type_used = r_list_new ();
+	r_list_foreach (list, iter, var) {
+		r_list_append (type_used, var->type);
+	}
+	RList *uniq = r_list_uniq (type_used, typecmp);
 	r_list_free (type_used);
 	return uniq;
 }
+
+static void enum_type_fini(void *e, void *user) {
+	(void)user;
+	RAnalEnumCase *cas = e;
+	free ((char *)cas->name);
+}
+
+static RAnalBaseType *get_enum_type(RAnal *anal, const char *sanitized_name) {
+	r_return_val_if_fail (sanitized_name != NULL, NULL);
+
+	RAnalBaseType *base_type = malloc (sizeof (RAnalBaseType));
+	if (!base_type) {
+		return NULL;
+	}
+	base_type->kind = R_ANAL_BASE_TYPE_KIND_ENUM;
+	
+	RAnalBaseTypeEnum base_enum;
+	char *key = r_str_newf ("enum.%s", sanitized_name);
+	if (!key) {
+		free (base_type);
+		return NULL;
+	}
+	char *members = sdb_const_get (anal->sdb_types, key, NULL);
+	free (key);
+
+	RVector cases;
+	r_vector_init (&cases, sizeof (RAnalEnumCase), enum_type_fini, NULL);
+
+	if (!r_vector_reserve (&cases, (size_t) sdb_alen (members))) {
+		free (base_type);
+		r_vector_free (&cases);
+		return NULL;
+	}
+
+	char *cur;
+	sdb_aforeach (cur, members) {
+		char *val_key = r_str_newf ("enum.%s.%s", sanitized_name, cur);
+		char *value = sdb_const_get (anal->sdb_types, val_key, NULL);
+		if (!val_key) {
+			free (base_type);
+			r_vector_fini (&cases);
+			return NULL;
+		}
+		RAnalEnumCase cas = {.name = strdup (cur), .val = strtol (value, NULL, 16)};
+
+		free (val_key);
+		
+		void *element = r_vector_push (&cases, &cas); // returns null if no space available
+		if (!element) {
+			free (base_type);
+			r_vector_fini (&cases);
+			return NULL;
+		}
+
+		sdb_aforeach_next (cur);
+	}
+
+	base_enum.cases = cases;
+	base_type->enum_data = base_enum;
+
+	return base_type;
+}
+
+// returns NULL if name is not found or any failure happened
+R_API RAnalBaseType *r_anal_get_base_type(RAnal *anal, const char *name) {
+	r_return_val_if_fail (name != NULL, NULL);
+	r_return_val_if_fail (anal != NULL, NULL);
+
+	char *name_sanitized = r_str_sanitize_sdb_key (name);
+
+	char *type = sdb_const_get (anal->sdb_types, name_sanitized, NULL);
+
+	// Right now just types: struct, enum, union are supported
+	if (!type || !(strcmp (type, "enum") || strcmp (type, "struct") || strcmp (type, "union"))) {
+		free(name_sanitized);
+		return NULL;
+	}
+	// Taking advantage that all 3 types start with distinct letter
+	// because the strcmp condition guarantees that only those will get to this flow
+	RAnalBaseType *base_type;
+	
+	switch (type[0]) {
+	case 's': // struct
+		break;
+	case 'e': // enum
+		base_type = get_enum_type (anal, name_sanitized);
+		// DEBUG print, TODO not forget to remove
+		r_cons_printf ("BaseType: %d\n", base_type->kind);
+		RAnalEnumCase *it;
+		r_cons_printf("Cases:\n");
+		r_vector_foreach (&base_type->enum_data.cases, it) {
+			r_cons_printf ("\tname: %s, value: %d\n", it->name, it->val);
```

Comment:
dont use tabs in output it will break visual panels

Reply:
It's just temporary print to see the function result for development